### PR TITLE
Set QA Agent as the default agent in SlackApp and add configuration for the obsolete agent

### LINKED ISSUE #261: [FEATURE] Switch the default agent to QA agent
**PROBLEM**
Now we have a relatively stable QA agent with many new features exclusively added to it. We should switch the default agent for the slack app to it :)

### PR DESCRIPTION
**Pull request by 20001LastOrder**
# Description
Resolves #261 

Now the QA agent will be used as the default agent. The `--obsolete` flag can be used to switch back to the old task agent. 
Also fixes an error in the citation validation that fails the tests when switching to the QA agent. 

# Your checklist for this pull request
Thank you for submitting a pull request! To speed up the review process, please follow this checklist:

- [x] My Pull Request is small and focused on one topic so it can be reviewed easily 
- [x] My code follows the style guidelines of this project (`make format`)
- [x] Commit messages are detailed
- [x] I have performed a self-review of my code
- [x] I commented hard-to-understand parts of my code
- [x] I updated the documentation (docstrings, `/docs`)
- [x] My changes generate no new warnings (or explain any new warnings and why they're ok)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All tests pass when I run `pytest tests` (offline mode)

Additional steps for code with networking dependencies (Does not apply):

- [ ] I followed the [offline and online testing guidelines](https://sherpa-ai.readthedocs.io/en/latest/How_To/Test/test.html#offline-and-online-testing)
- [ ] All tests pass when I run `pytest tests --external_api` (online mode, making network calls)

### CODE REVIEW

**Comment by oshoma** [comment]
Hey @20001LastOrder when I don't have Pinecone keys defined in `.env` I get this test failure. Can you resolve?

```python
_______________________________________________________________________________________________ test_extract_github_readme_with_valid_url_succeeds _______________________________________________________________________________________________

    def test_extract_github_readme_with_valid_url_succeeds():
        repo_url = "https://github.com/Aggregate-Intellect/sherpa"
>       content = extract_github_readme(repo_url)

tests/unit_tests/scrape/test_extract_github_readme.py:6: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sherpa_ai/scrape/extract_github_readme.py:58: in extract_github_readme
    save_to_pine_cone(content, metadata)
sherpa_ai/scrape/extract_github_readme.py:69: in save_to_pine_cone
    vectorstore = ConversationStore("Github_data", index, embeddings, "text")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <sherpa_ai.connectors.vectorstores.ConversationStore object at 0x7fa9ddf25100>, namespace = 'Github_data', db = <pinecone.index.Index object at 0x7fa9daa59b20>
embeddings = OpenAIEmbeddings(client=<class 'openai.api_resources.embedding.Embedding'>, async_client=None, model='text-embedding-a...show_progress_bar=False, model_kwargs={}, skip_empty=False, default_headers=None, default_query=None, http_client=None)
text_key = 'text'

    def __init__(self, namespace, db, embeddings, text_key):
        self.db = db
        self.namespace = namespace
>       self.embeddings = embeddings
E       AttributeError: can't set attribute

sherpa_ai/connectors/vectorstores.py:24: AttributeError
```

**Comment by oshoma** [comment]
Re:   AttributeError: can't set attribute

This bug still occurs when I have pinecone set up. Maybe you're using a different langchain version?

```
(base) ➜  src git:(slackapp) ✗ poetry show langchain
 name         : langchain                                             
 version      : 0.0.332                                               
 description  : Building applications with LLMs through composability 

dependencies
 - aiohttp >=3.8.3,<4.0.0
 - anyio <4.0
 - async-timeout >=4.0.0,<5.0.0
 - dataclasses-json >=0.5.7,<0.7
 - jsonpatch >=1.33,<2.0
 - langsmith >=0.0.52,<0.1.0
 - numpy >=1,<2
 - pydantic >=1,<3
 - PyYAML >=5.3
 - requests >=2,<3
 - SQLAlchemy >=1.4,<3
 - tenacity >=8.1.0,<9.0.0
```

Hmm... I also see the test will probably fail if I don't set up Pinecone with an index named "langchain".

We should change the test so that it's more robust
- skipped when Pinecone is not set up 
- skipped when OpenAI key is not set
- make this an external API test with mocks?

File: src/apps/slackapp/tests/test_switch_agents.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+from unittest.mock import MagicMock, create_autospec, patch
+
+from langchain.llms.fake import FakeListLLM
+from slackapp.bolt_app import get_response
+
+
+@patch("sherpa_ai.agents.qa_agent.QAAgent.run")
+@patch("sherpa_ai.task_agent.TaskAgent.run")
+def test_normal_mode(mock_task_agent_run, mock_qa_agent_run):
```
Comment (oshoma) [inline src/apps/slackapp/tests/test_switch_agents.py:9]: test_normal_mode_uses_qa_agent

File: src/apps/slackapp/tests/test_switch_agents.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        llm=llm,
+    )
+
+    mock_qa_agent_run.assert_called()
+    mock_task_agent_run.assert_not_called()
+
+
+@patch("sherpa_ai.agents.qa_agent.QAAgent.run")
+@patch("sherpa_ai.task_agent.TaskAgent.run")
+def test_obsolete_mode(mock_task_agent_run, mock_qa_agent_run):
```
Comment (oshoma) [inline src/apps/slackapp/tests/test_switch_agents.py:30]: test_obsolete_mode_uses_task_agent

File: src/sherpa_ai/config/task_config.py [hunk 3]
```diff
@@ -71,6 +72,12 @@ def from_config(cls, configs: List[str]) -> "AgentConfig":
             help="enable performing the reflection step for each agent.",
         )
 
+        parser.add_argument(
+            "--obsolete",
+            action="store_true",
+            help="enable performing the obsolete step for each agent.",
```
Comment (oshoma) [inline src/sherpa_ai/config/task_config.py:78]: hmm... maybe "enable use of task agent (obsolete)"

File: src/sherpa_ai/output_parsers/link_parse.py [hunk 4]
```diff
@@ -36,14 +36,14 @@ def replace_with_symbol(match: re.Match):
         def replace_with_link(match: re.Match):
             logger.debug(match)
             doc_id = int(match.group(1))
-            if doc_id < 0 or doc_id >= len(self.links):
+            if doc_id <= 0 or doc_id > len(self.links):
```
Comment (oshoma) [inline src/sherpa_ai/output_parsers/link_parse.py:39]: If you fixed a bug here then add a test

File: src/apps/slackapp/slackapp/bolt_app.py [hunk 5]
```diff
@@ -103,30 +107,53 @@ def get_response(
...
+    if not agent_config.verbose:
+        verbose_logger = DummyVerboseLogger()
 
     memory = get_vectordb()
+    tools = get_tools(memory, agent_config)
 
-    question, agent_config = AgentConfig.from_input(question)
-    verbose_logger = verbose_logger if agent_config.verbose else DummyVerboseLogger()
+    if agent_config.obsolete:
+        verbose_logger.log("⚠️🤖 Use obsolete mode ...")
```
Comment (oshoma) [inline src/apps/slackapp/slackapp/bolt_app.py:123]: Use task agent (obsolete)...

File: src/sherpa_ai/config/task_config.py [hunk 6]
```diff
@@ -11,6 +11,7 @@ class AgentConfig(BaseModel):
     verbose: bool = True
     gsite: list[str] = []
     do_reflect: bool = False
+    obsolete: bool = False
```
Comment (oshoma) [inline src/sherpa_ai/config/task_config.py:14]: General comment: for anyone who hasn't seen this code before, "obsolete" is too vague.  I suggest calling it something like `use_task_agent`, and in the parser.add_argument and/or elsewhere in the docs you can explain this mode of operation is obsolete.

**Review by oshoma — CHANGES_REQUESTED** [review]
Suggesting some aesthetic changes here and maybe adding a test or two.

See also the bug I ran into... I haven't been able to get that broken test to succeed. Not sure whether it's related.

**Comment by 20001LastOrder** [comment]
> Re: AttributeError: can't set attribute
> 
> This bug still occurs when I have pinecone set up. Maybe you're using a different langchain version?
> 
> ```
> (base) ➜  src git:(slackapp) ✗ poetry show langchain
>  name         : langchain                                             
>  version      : 0.0.332                                               
>  description  : Building applications with LLMs through composability 
> 
> dependencies
>  - aiohttp >=3.8.3,<4.0.0
>  - anyio <4.0
>  - async-timeout >=4.0.0,<5.0.0
>  - dataclasses-json >=0.5.7,<0.7
>  - jsonpatch >=1.33,<2.0
>  - langsmith >=0.0.52,<0.1.0
>  - numpy >=1,<2
>  - pydantic >=1,<3
>  - PyYAML >=5.3
>  - requests >=2,<3
>  - SQLAlchemy >=1.4,<3
>  - tenacity >=8.1.0,<9.0.0
> ```
> 
> Hmm... I also see the test will probably fail if I don't set up Pinecone with an index named "langchain".
> 
> We should change the test so that it's more robust
> 
> * skipped when Pinecone is not set up
> * skipped when OpenAI key is not set
> * make this an external API test with mocks?

We are aware of this bug, see this issue: #256

**Review by oshoma — APPROVED** [review]

**Pull request merged**